### PR TITLE
Fix optimizer cuda health check graph break (can be done in the compiler)

### PR DESCRIPTION
**Pull request by mlazos**
- Ignore the health check if we are compiling
- Don't disable the function anymore

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
Does this mean the "capturable" flag will be unnecessary in torch.compile land because everything will already be capturable?

**Review by janeyx99 — COMMENTED** [review]
Does this mean the "capturable" flag will be unnecessary in torch.compile land because everything will already be capturable?

**Comment by mlazos** [comment]
> Does this mean the "capturable" flag will be unnecessary in torch.compile land because everything will already be capturable?

Yeah, but not entirely because the compiler cannot trace python math. The different variants of adam will probably have to have some if is_compiling() to use pytorch ops for the python math. That will be a in a follow up PR though where we can discuss.

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def __repr__(self):
 
     # Currently needed by Adam and AdamW
     def _cuda_graph_capture_health_check(self):
-        if torch.has_cuda and torch.cuda.is_available():
+        if not is_compiling() and torch.has_cuda and torch.cuda.is_available():
```
Comment (ngimel) [inline torch/optim/optimizer.py:229]: you don't correctly set `capturing` status when you are compiling? What happens then?

Comment (mlazos) [reply torch/optim/optimizer.py:229]: So if we're compiling I think I'm going to take the capturable path automatically vs setting the flag

Comment (ngimel) [reply torch/optim/optimizer.py:229]: Leave a comment about that

**Review by albanD — APPROVED** [review]
Sure

**Pull request closed without merging**